Filter instance names with regex

StevenMaude opened the pull request:
Addresses #1.

Use a regex at the command line with the `-filter` flag to restrict the list of instances to those that match the regex.

The watch behaviour previously accessed via `@` now works with the `-watch` flag.

One issue is that the table formatting looks a bit odd as the number of instances displayed may differ from those that are used in `goUp()`. It would be possible to filter the instances at that point. What I'm not sure about is whether the number used to select a particular instance in a filtered list of instances should be consistent with the number in the unfiltered list.

Feedback also welcome on:
- should `@` still be accessible via `@`, or is it fine with `-watch`?
- is a regex a sensible way to do the filtering? (There's no easy negation)
- anything bad I've done as I'm still trying to pick up Go.

pwaller commented on main.go:
```diff
@@ -23,6 +25,19 @@ func ShowInstances(instances []*Instance) {
+			regex, err := regexp.Compile(*filterRegex)
+			if err != nil {
+				fmt.Println(*filterRegex)
+				log.Fatal("Bad regex filter.")
```
In general the whole PR looks good. Only comment is on this line. As a rule, in go, you really want to show the user the value of `err` whenever possible. Otherwise, they might not be able to tell what's wrong. The usual way I would achieve this is as so:

``` golang
log.Fatalf("Regex does not compile: %v", err)
```

Or maybe

```
log.Fatalf("Failed to compile regex %q: %v", *filterRegex, err)
```

A good error message makes it clear what failed and why. `Bad regex filter` leaves it ambiguous what the program was trying to do with the regex when it failed.

pwaller commented:
Ping @StevenMaude

StevenMaude commented:
Haven't forgotten about this; will tidy it in a spare moment this week.

StevenMaude committed:
Initial attempt at instance filtering

StevenMaude committed:
Ignore watch and filter arguments for SSH

A nicer solution might be to explicitly specify SSH arguments with other
command line arguments.

StevenMaude committed:
Filter when we create the slice of instances

Instead of at the point of displaying them. This makes it easier to
figure out the correct number of instances, which fixes table redrawing.

(len(instances) wouldn't necessarily reflect the number of instances
displayed, meaning the cursor may move up the wrong amount.)

StevenMaude commented on util.go:
```diff
@@ -4,12 +4,19 @@ import (
+			strings.HasPrefix(param, "-filter") ||
+			strings.HasPrefix(param, "--filter") {
+			continue
+		}
```
Not sure if there's a better way of doing this. The problem is that the jump argument keywords are otherwise taken as SSH arguments which then breaks the SSH command as it tries to do e.g. `ssh -filter`.

pwaller replied:
It's not obvious to me how this works since surely that just results in the argument to those parameters being passed to SSH? What happens with `--filter foo`?

I guess this is a bigger problem perhaps outside the scope of this PR.

StevenMaude commented on instance.go:
```diff
@@ -86,8 +88,27 @@ func InstancesFromEC2Result(in *ec2.DescribeInstancesOutput) []*Instance {
+		}
+		filteredOut = append(filteredOut, i)
+	}
+
```
Looping again just to filter the unwanted instances seems redundant as we're looping over in just above. However, there didn't seem to be an easy way to get the instance name for matching at that point.

StevenMaude commented:
Delayed on this, but I think it's almost there.

I addressed the logging issue mentioned above, fixed the cursor position on table redraw and actually made the filtering usable; the `jump` command line arguments got in the way of the `ssh` arguments.

StevenMaude commented:
What I don't know how to best to deal with is the now broken `ssh` arguments; you get `flag provided but not defined` if you try and use them.

Guess we could make these a flag too, e.g. `-ssh-before-host` and `-ssh-after-host`.

pwaller commented:
Ping @StevenMaude sorry I broke this PR, can you rebase at some point?

pwaller commented:
Ping @StevenMaude is this still needed with #7 ?

StevenMaude commented:
Probably not; may be something you want in future if you happen to have a lot of instance in a single VPC, but that fixes it for me.
